File: 2903-insert-greatest-common-divisors-in-linked-list/2903-insert-greatest-common-divisors-in-linked-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
function gcd(a, b) {
    if(b == 0) return a;
    return gcd(b, a % b);
}

var insertGreatestCommonDivisors = function(head) {
    let curr = head;
    
    while(curr.next != null) {
        const val = gcd(curr.val, curr.next.val);
        let temp = curr.next;
        let newNode = new ListNode(val);
        curr.next = newNode;
        newNode.next = temp;
        curr = temp;
    }

    return head;
};

File: 2903-insert-greatest-common-divisors-in-linked-list/2903-insert-greatest-common-divisors-in-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* insertGreatestCommonDivisors(ListNode* head) {
        ListNode* curr = head;
        while(curr -> next) {
            int val = __gcd(curr -> val, curr -> next -> val);
            ListNode* temp = curr -> next;
            ListNode* newNode = new ListNode(val);
            curr -> next = newNode;
            newNode -> next = temp;
            curr = temp;
        }

        return head;
    }
};